### PR TITLE
Implement improved list views for categories, first page and programs.

### PR DESCRIPTION
**Pull request by mlehtima**
Work in progress. Any comments welcome.

### CODE REVIEW

**Comment by mlehtima** [comment]
Should the search page also use same list view delegate. The delegate could be separated to a file.

File: qml/js/yleApi.js [hunk 1]
```diff
@@ -28,12 +28,57 @@ var filterAvailablePrograms = function(programs) {
     }, []);
 }
 
+var zeropad = function(str, size) {
+    var s = str;
+    while (s.length < size) s = "0" + s;
```
Comment (joonne) [inline qml/js/yleApi.js]: I'd prefer always to use curly braces after while, if.

Objections? :)

File: qml/js/yleApi.js [hunk 2]
```diff
@@ -28,12 +28,57 @@ var filterAvailablePrograms = function(programs) {
...
 }
 
+var zeropad = function(str, size) {
+    var s = str;
+    while (s.length < size) s = "0" + s;
+    return s;
+}
+
+var parseDuration = function(dur) {//PT1H58M45S
+    if (!dur)
```
Comment (joonne) [inline qml/js/yleApi.js]: same

File: qml/js/yleApi.js [hunk 3]
```diff
@@ -28,12 +28,57 @@ var filterAvailablePrograms = function(programs) {
...
+        return "";
+    var d = dur.substr(2);
```
Comment (joonne) [inline qml/js/yleApi.js]: I'd prefer more descriptive (longer) variable names

File: qml/js/yleApi.js [hunk 4]
```diff
@@ -28,12 +28,57 @@ var filterAvailablePrograms = function(programs) {
...
+
+var parseTime = function(str) {//2014-01-23T21:00:07+02:00
+    var s = Qt.formatDateTime(new Date(str), "ddd dd.MM hh:mm")
+    console.log(str+"->"+s)
+    return s;
+}
+
+function arraySearch(arr, key, sub, val) {
+    for (var i = 0; i < arr.count; i++) {
+        if (sub && arr.get(i)[sub] && arr.get(i)[sub][key] === val)
```
Comment (joonne) [inline qml/js/yleApi.js]: same

File: qml/js/yleApi.js [hunk 5]
```diff
@@ -28,12 +28,57 @@ var filterAvailablePrograms = function(programs) {
...
+            return i;
+        if (arr.get(i)[key] === val)
```
Comment (joonne) [inline qml/js/yleApi.js]: same

File: qml/js/yleApi.js (hunk 4 above)
Comment (joonne) [inline qml/js/yleApi.js]: Where are `.get` and `.count` functions coming from?

File: qml/js/yleApi.js [hunk 6]
```diff
@@ -28,12 +28,57 @@ var filterAvailablePrograms = function(programs) {
...
+            return i;
+    }
+//    console.log("startTime not found")
+    return -1;
+}
+
+var findPubTime = function(arr) {
+    if (!arr)
```
Comment (joonne) [inline qml/js/yleApi.js]: same

File: qml/js/yleApi.js [hunk 7]
```diff
@@ -58,7 +103,7 @@ function getProgramById(id) {
 }
 
 function getProgramsByCategoryId(categoryId, limit, offset) {
-    var url = apiUrl + programsUrl + '&category=' + categoryId + '&availability=ondemand&mediaobject=video';
+    var url = apiUrl + programsUrl + '&category=' + categoryId + '&availability=ondemand&mediaobject=video&offset='+offset+'&limit='+limit;
```
Comment (joonne) [inline qml/js/yleApi.js:103]: cool!

File: qml/pages/ProgramsPage.qml [hunk 8]
```diff
@@ -7,10 +7,24 @@ Page {
     id: page
 
     property var category: ({})
+    property int offset: 0
+    property int limit: 25
 
     Component.onCompleted: {
```
Comment (joonne) [inline qml/pages/ProgramsPage.qml]: This can be replaced now with `Component.onCompleted: getPrograms()` ?

File: qml/pages/ProgramsPage.qml [hunk 9]
```diff
@@ -31,29 +45,100 @@ Page {
         header: PageHeader {
             title: qsTr("Programs")
         }
+
+        PullDownMenu {
```
Comment (joonne) [inline qml/pages/ProgramsPage.qml]: :+1:

**Review by joonne — COMMENTED** [review]
Looks good! Programs page is way better and the pagination is great.

Delegate could indeed be separated into own component.

**Comment by joonne** [comment]
I opened that Trello like board in the Projects section, would you feel comfortable using that to track what we are doing? Just need to add issue / PR to the Project Näyttämö so it is semi automated.

File: qml/js/yleApi.js (hunk 1 above)
Comment (mlehtima) [reply qml/js/yleApi.js]: I actually also prefer that way, this was taken from my old work-in-progress code so I didn't yet check all of the formatting.

File: qml/js/yleApi.js (hunk 4 above)
Comment (mlehtima) [reply qml/js/yleApi.js]: Those should be basic javascript array functions.

File: qml/js/yleApi.js (hunk 1 above)
Comment (joonne) [reply qml/js/yleApi.js]: :+1:

File: qml/js/yleApi.js (hunk 4 above)
Comment (joonne) [reply qml/js/yleApi.js]: I might be confused also but I'm pretty sure they aren't, they might result from Array / ListModel conversions, but I did not see any explicit conversions happening anywhere.

Should check with typeof & console logs what kind of object it actually is :)

**Comment by mlehtima** [comment]
Updated code based on comments and fixed some other issues I noticed.

**Comment by joonne** [comment]
Good!

Did you btw modify the finnish translations by hand? You can ofc if you like but translations are hosted on https://www.transifex.com/joonne/nayttamo/dashboard/ also and it is quite a bit simpler there.

File: rpm/harbour-nayttamo.spec [hunk 10]
```diff
@@ -65,6 +65,7 @@ desktop-file-install --delete-original       \
 %defattr(-,root,root,-)
 %{_bindir}
 %{_datadir}/%{name}
+%{_datadir}/%{name}/translations
```
Comment (joonne) [inline rpm/harbour-nayttamo.spec]: did Qt Creator modify these or was there a bug? I'm a bit confused with these two files still... I've only used them to increase version numbers so far :sweat_smile:

**Review by joonne — APPROVED** [review]

**Comment by joonne** [comment]
ready to merge?

I sometimes use the convention of prepending PR name with `WIP:` when feature is available for comments but not finished, could we use that here also?

File: rpm/harbour-nayttamo.spec (hunk 10 above)
Comment (mlehtima) [reply rpm/harbour-nayttamo.spec]: This change is actually not needed. I had some issues with translations until I figured out that running the app from Sailfish SDK deploy makes translations fail and this remained from my testting. I can remove this if needed.

**Comment by mlehtima** [comment]
I did edit the translations manually because I forgot about transifex. Next time I will use transifex. Depending on that last comment this can be merged. I can make the trivial change if needed. Also next time I will use WIP in the title to make things more clear.

**Comment by joonne** [comment]
If you could undo the unneeded change, that would be great :) I'll merge right away then.

**Comment by mlehtima** [comment]
Removed the unneeded changes.

**Pull request merged**